TASK: Refactor and optimize of session data storage

mficzel opened the pull request:
Before the SessionMetadata and SessionData was written with every request which caused network traffic, storage wear and also made race conditions much more likely when parallel requests changed session data. 

In total this can reduce the number of write operations on the Session caches by 80-90% removing storage and network load as those caches always are persistent and shared across clusters.

1. The improvements are on top of the Neos 9 already reducing the Flow_Session_Storage write load by not always storing the "lastVisitedNode" in the session.
2. The improvements mostly occur after sending the result "onShutdown" so this will not improve single requests but overall performance and number of parallel requests.

## 1. SessionMetadataStore / Flow_Session_MetaData

### Problem 

The session metadata (SessionId, StorageId, LastActivity .. ) is usually written at shutdown of every single request to the session metadata cache even when nothing changed. 

### Optimization 

This is optimized by the new setting `Neos.Flow.session.updateMetadataThreshold` that allows to configure the interval for updating the session metadata when nothing but the `lastActivityTimestamp` has changed. This removes lots of cache writes and avoids network traffic or storage wear. The session metadata is also converted to a single value object that combines SessionID, StorageID and lastActivityTimestamp.

## 2. SessionKeyValueStore / Flow_Session_Storage: 

### Problem

Session data is written to the Flow_Session_Storage cache once Session->putData is called. In case of flow this mostly is the Objects of scope `@Flow\Scope("session")` that are stored on shutdown. Those objects are sometimes modified but during most requests nothing changes here and at the end of the request the same data is added to the cache again with another redundant cache write. 

### Optimization

The SessionDataStore optimizes this be keeping a hash of the value all previously read keys and avoids writes if the serialized content that is stored yields the same hash. That way only once session-data was actually changed the session objects are actually written to the cache. This also lessens the probability of some race conditions drastically that can occur when multiple parallel requests work on the same session.

The following redundant behavior was also removed:
- All session metadata records in the cache were previously tagged with `session` for iterating over them again. This is replaced by the `retrieveAll` method.
- The current authentication providers were always stored in the session data as `Neos_Flow_Security_Accounts ` but were unused  
- Do we want to release this as 8.4 or 9.0? In case of 9.0 the SessionMetaData ValueObject will be adjusted to php 8.1 style.

Resolves: https://github.com/neos/flow-development-collection/issues/525

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

Some questions during the review could be:
- Are there better ways to determine written objects are modified than comparing hashes of serialized values?: I did not find one.
- Should the comparison of written data with hashes of existing data be implemented in the cache frontend instead?: I think this would consume to much memory we have lots of cache items.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mficzel committed:
TASK: Extract Flow_Session_MetaData and Flow_Session_Storage caches from `Session` and `SessionManager` into `SessionDataStore` and `SessionMetaDataStore`

This is the first step of refactoring the storage of session data to value objects.

kitsunet commented:
Great idea <3

mficzel committed:
TASK: Introduce SessionMetaData value object to replace the previous array based metadata

mficzel committed:
TASK: Store SessionIdentifier in SessionMetaData

This will enable to have a single record to contain the whole metadata for the session.

mficzel committed:
TASK: Store single sessionMetaData object in session instead of multiple properties

mficzel committed:
TASK: Use seperate factory methods `create` and `createRemote` for normal and remote sesions

mficzel committed:
TASK: Adjust session data store to handle the entryIdentifier generation internally and accept session metadata externally

mficzel committed:
TASK: Remove storing of account informations in the session which seems to be unused

mficzel committed:
TASK: Remove deprecated collect garbage method from session

The method collectGarbage is now located in the sessionManagerInterface.

mficzel committed:
TASK: Make style ci happy

mficzel committed:
TASK: Remove session metadata after destroy

some tests had to be adjusted for that

mficzel committed:
TASK: Move flow session object storage from sessionDataStore to session

mficzel committed:
TASK: Add debounce hash to session data store and use string frontend

By serializing in SessionDataStore and using the string frontend a debounce hash can be held to
avoid writing the same data again even when it was already written.

mficzel committed:
FEARURE: Add `updateMetadataThreshold` to session settings

The metadata store will keep a copy of all metadata entries while reading and will only update metadata entries that
have been modified or that are older than the given threshold.

mficzel commented:
Keep this as draft for now until we decide wether this should go to 9.0 or a hypothetical 8.4. In case of 9.0 would refactor the SessionMetaData Value object to PHP 8.1 style.

fcool commented:
Yay! The idea is great - but I fear this won't work with Entites used in sessions - as we reverted my change, that serializiation becomes repeatable.

So if one uses serialize on an object used in session, the session will again "always be modified".

But before seeing only the troubles... I'll try to prove my assumption based on this PR.

The general idea and implementation are looking VEEERY neat

mficzel commented:
@fcool as i understand you are referring to #2009 .. while this is valid you would have to either call serialize twice for some reason or have this as an argument and async jobqueue call somewhere. Sounds like an edge case to me, do i understand that correct?

Sidenote: I personally would avoid entities in methods calls that are asyncronously handled by the jobqueue and only use identifiers as arguments there.

fcool commented:
Exactly. 
But no, "one" would only need to call it once. The session would be "twice". 

So if you put your objects, which are stored in session for whatever reason in a cache or use the serialize call for other reasons, the session - the "second" call, will not get the expected result.

And its no "edge case". Neos itself does it (storing the Assets to display or already displayed in session for example)

mficzel commented:
Will try to do some benchmark to see wether this prevents this from beeing effective at all or wether it only is not always effective.  In general it should still work as before but may not be able to optimize in certain cases.

mficzel committed:
TASK: Read sessuion data store should read before writing to avoid constant overwriting with same values

mficzel commented:
Just took some totally non scientific numbers during a Neos session:  Login to Neos, change a text on the homepage and another page, open 4 other documents and log out again.

- Before: 
  - Flow_Session_MetaData: get 11 , set 33
  - Flow_Session_Storage: get 26, set 23
- After: 
  - Flow_Session_MetaData: get 10 , set 3
  - Flow_Session_Storage: get 20, set 13

That looks like at least 50% less many writes to the Flow_Session_Storage and 90% less to Flow_Session_MetaData.  And that with backend login and logout which has quite some legitimate stuff to write. I would expect the numbers to be even higher outside of the Neos Backend as this stores the last visited node all the time.

mficzel commented:
@fcool Tested a bit and had to adjust the code a little for writes when no read was performed before (lastVisitedNode). To me it seems to work at least as good as before with quite significantly reduced writes. If there are still some edge cases left this should not make the situation for anyone worse so imho this still seems to be the right direction.

PS: Maybe we should discourage to use entities in session data and jobqueue calls in favor of plain value objects and identifiers ... seems a good practice anyways.

Sebobo commented:
Sadly cannot give any good feedback on this as I never looked into the Session related code. But sounds great!
Or is there something I could test with a large project and check for differences?

mficzel commented:
@Sebobo as you have much experience with optimizing caches you might be able to measure compare the performance impact especially for larger projects. Also it would be good to compare impact for FlowApplications (with Sessions) that are not the Neos Backend.

PS: We had flow sessions kill a bigger galera cluster because of the current non optimized writing behavior. So this is not purely theoretical.

mficzel committed:
TASK: Refactor session metadata to php 8 value objects

The properties get the readonly flag after we rebase / merge to php 8.1

mhsdesign commented on Neos.Flow/Classes/Session/Data/SessionMetaData.php:
```diff
@@ -0,0 +1,140 @@
+        public /** readonly */ string $sessionIdentifier,
+        public /** readonly */ string $storageIdentifier,
+        public /** readonly */ int $lastActivityTimestamp,
+        public /** readonly */ array $tags
```
.

mhsdesign commented on Neos.Flow/Classes/Session/Data/SessionMetaData.php:
```diff
@@ -0,0 +1,140 @@
+            return false;
+        }
+
+        if (array_diff($this->tags, $other->tags) || array_diff($other->tags, $this->tags)) {
```
this seems not super optimised ...

mhsdesign replied:
adjusted via. https://github.com/neos/flow-development-collection/pull/3254/commits/1a16e7e075d9ad3e3320a13da73ddf25355e18f4

mficzel committed:
TASK: Adjust comparison operators in SessionMetadata

mficzel committed:
TASK: Rename SessionDataStore to SessionKeyValueStore and mark both session stores internal

mficzel committed:
TASK: Introduce StorageIdentifier value object

mficzel committed:
TASK: Introduce SessionIdentifier value object

mficzel commented:
close and reopen to trigger tests

mficzel committed:
TASK: Make linter happy

mficzel committed:
TASK: Remove weirdness that is not needed in neos 9

Since the odd behavior of writing "lastVisitedNode" to the session cache without reading it first is gone.

mficzel commented:
I rebased this on Flow 9 and created fresh stats. 

This pr yields a 85% resp. 90% improvement in write (set) cache operations for the session caches in this arguably simple scenario while all the other call numbers remain the same.

procedure: Login * refresh homepage 10x * logout

before:
```
+---------------------------------+-------+
| key                             | num   |
+---------------------------------+-------+
| Flow_Session_MetaData_has       | 23    |
| Flow_Session_MetaData_get_num   | 47    |
| Flow_Session_MetaData_set_num   | 25    |
| Flow_Session_MetaData_remove    | 2     |
| Flow_Session_Storage_has        | 24    |
| Flow_Session_Storage_get_num    | 24    |
| Flow_Session_Storage_set_num    | 47    |
+---------------------------------+-------+
```

after:
```
+---------------------------------+-------+
| key                             | num   |
+---------------------------------+-------+
| Flow_Session_MetaData_has       | 23    |
| Flow_Session_MetaData_get_num   | 48    |
| Flow_Session_MetaData_set_num   | 4     |
| Flow_Session_MetaData_remove    | 2     |
| Flow_Session_Storage_has        | 25    |
| Flow_Session_Storage_get_num    | 24    |
| Flow_Session_Storage_set_num    | 5     |
+---------------------------------+-------+
```

mhsdesign commented on Neos.Flow/Classes/Session/Data/SessionIdentifier.php:
```diff
@@ -0,0 +1,44 @@
+class SessionIdentifier
+{
+    private function __construct(
+        public /*readonly*/ $value
```
can be _real_ readonly now with Flow 9

mficzel replied:
You are right, i forgot about that. Will go over again and also add proper argument and return types to the whole session classes.

mficzel replied:
is adjusted

mhsdesign reviewed (commented):
Thank you already for the hard work :D
Fyi for other reviewers: It might not seem like it but i reviewed the change already together with Wilhelm and Martin so the last commits _were_ already the adjustments:

- `TASK: Adjust comparison operators in SessionMetadata`
- `TASK: Rename SessionDataStore to SessionKeyValueStore and mark both internal`
- `TASK: Introduce StorageIdentifier value object`
- `TASK: Introduce SessionIdentifier value object`

and we also discussed that the optimisation sorely in the intention for `lastVisitedNode` will not be needed for 9.0 so thanks for following up on that and reverting this `TASK: Remove weirdness that is not needed in neos 9` again ❤️ 

I found that the properties can now be finally _real_ readonly properties and i would like to have a simple comment for `StorageIdentifier` `SessionIdentifier` what they represent (for us flow devs). Besides these nitpicks i will look into the adjustments above you made to refresh my memory and give you my approval soonish.

mficzel committed:
TASK: Upgrade to php 8 readonly properties

After rebasing we have php 8.1 as min version and can rely on that.

mhsdesign commented on Neos.Flow/Classes/Session/Data/StorageIdentifier.php:
```diff
@@ -0,0 +1,44 @@
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Utility\Algorithms;
+
+/**
```
i would like to have a simple comment for StorageIdentifier SessionIdentifier what they represent (for us flow devs). 

After your last explanation i already forgot about their meaning a little 😅 as im not that deep into these things ^^

mficzel replied:
Added comments to explain as good as i can ... the distinction between session and storage id has been there beforehand and while i see it i can only guess the real reason and or necessity.

mhsdesign replied:
Thank youuuu!!

mficzel committed:
DOCS: Add comments to hopefully explain the distinction between session and storage identifier

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/sessionStorageRefactoring

mhsdesign reviewed (approved):
Im not really sure that we still have to upcast old sessions from 8.3 or 9.0-dev previously via `createFromSessionIdentifierStringAndOldArrayCacheFormat` but as the logic is now in place and to be marked as depreacted and to be removed with Flow 10 we can surely keep it.

Thanks for putting the work into it and also for adjusting the code as requested ❤️

robertlemke reviewed (approved):
Thanks a lot

This looks good to me, makes sense and did work for me using some application I tested it with.
